### ISSUE TITLE
`torch.compile` advanced index on 3-D float16 tensor causes PCIe bus fence and process abort on Spyre

### ISSUE THREAD

**Issue by KanakaPathivada** [torch-spyre/torch-spyre#4724]
### What happened?

Using `torch.compile` with the Inductor backend, a function that performs advanced row indexing (`w[ids]`) on a 3-D `float16` tensor (`shape=(8, 64, 32)`), multiplies by a broadcast scalar vector, and reduces via `.sum(dim=0)` causes the Spyre card to trigger a **PCIe bus master fence**. The Python process then terminates with `SIGABRT` (signal 6).

The index tensor (`ids`) has dtype `int64`. The Spyre backend downcasts it to `int32` (with a warning) before dispatching the gather to the hardware. The card raises:

```
RAS::PCI::BusFence — The card encountered a problem and triggered a PCIe bus fence operation
```

The same function executes correctly on the CPU.

This is a process-fatal event. The session cannot be recovered after this error.

### What did you expect to happen?

`torch.compile` with advanced row indexing (`w[ids]`) should either:

1. Execute successfully on Spyre and return a numerically correct result matching CPU, or
2. Fall back gracefully with a Python-level `RuntimeError` or `NotImplementedError` if the operation is not fully supported on Spyre hardware.

It should **not** cause a PCIe bus fence or terminate the Python process with `SIGABRT`.

### How can we reproduce it?

```python
import torch

device = torch.device("spyre:0")
w      = torch.randn(8, 64, 32, dtype=torch.float16)
ids    = torch.randint(0, 8, (16,), dtype=torch.int64)
scores = torch.randn(16, dtype=torch.float16).softmax(dim=0)

def fn(table, i, s):
    return (table[i] * s[:, None, None]).sum(dim=0)

# CPU reference — executes correctly
ref = fn(w, ids, scores)

# Spyre compiled — triggers PCIe bus fence and aborts the process
got = torch.compile(fn, dynamic=False)(
    w.to(device), ids.to(device), scores.to(device)
).cpu()
torch.testing.assert_close(got, ref, atol=0.01, rtol=0.01)
```

**Observed result:** The process prints the int64 downcast warning then immediately aborts with `SIGABRT` / PCIe bus fence. There is no Python exception — the process is killed.

**Expected result:** Either a correct `(64, 32)` float16 tensor matching `ref`, or a clean Python-level error.

### Any environmental details we need to know?

| | |
|---|---|
| Platform | linux (x86\_64) |
| Python | 3.12.13 |
| PyTorch | 2.13.0+cpu |
| torch-spyre | commit `29c91d94edfa22813db61d4d958ede0ef56f3be2` |
| Backend | `torch.compile(backend="inductor")` |
| Device | `spyre:0` |
| Input dtype | `torch.float16` |
| Index dtype | `torch.int64` (downcast to `int32` by Spyre backend) |

### Root cause hypothesis

The `int64 → int32` downcast warning before the crash points to the index-handling path. The likely causes are one or more of:

1. **3-D gather address calculation is incorrect**: `w[ids]` where `w` is `(8, 64, 32)` requires computing a row offset of `ids[i] * 64 * 32` elements. If this is computed in `int32` after the downcast, intermediate products overflow for non-trivial shapes/offsets, producing an out-of-bounds hardware address.

2. **Fused kernel exceeds a hardware limit**: The fused graph (`gather → broadcast multiply → reduce`) may violate a hardware constraint (e.g. command buffer size, descriptor count, or DMA address range), causing the hardware watchdog to fence the bus.

3. **Index tensor not normalised before dispatch**: As with the negative-index case, the index values may be passed directly to the hardware gather without validation or normalization, and the hardware address calculation in the fused kernel produces an address outside the allocated memory region.

### Comparison table

| Execution path | Result |
|---|---|
| CPU + `torch.compile(backend="inductor")` | Pass — correct `(64, 32)` float16 result |
| Spyre eager `w[ids]` (3-D) | Not tested separately |
| Spyre + `torch.compile(backend="inductor")` | PCIe bus fence → `SIGABRT` — process killed |

### Relevant source locations

- `aten.index_select` in `COMPILED_OPS` (`eager.py` line 404) — the compiled gather path that reaches the hardware.
- `int64 → int32` downcast (`types_mapping.h:95`) — index type narrowing that may cause overflow in address arithmetic.

### Relevant log output

```Shell
/home/mahalakshmip/dt-inductor/torch-spyre/torch_spyre/_monkey_patch.py:177: UserWarning: Backend Spyre does not support int64; downcasting to int32 may change values outside the 32-bit range. You can silence this via warnings.filterwarnings(...) or spyre.set_downcast_warning(False) or TORCH_SPYRE_DOWNCAST_WARN env. variable. (Triggered internally at /home/mahalakshmip/dt-inductor/torch-spyre/torch_spyre/csrc/types_mapping.h:95.)
  res = orig_to(self, *args, **kwargs)
[unspecified]  ERRR 21.09.2026 12:28:21.710621 [                  ras_base.hpp:  74] {"action":"reset","category":"hardware","code":"0xa35e","description":"The card encountered a problem and triggered a PCIe bus fence operation","message":"PCIe bus master fence","name":"RAS::PCI::BusFence","severity":"ERROR","step":"Reset, reseat, or replace the card","threshold_action":"deconfigure","threshold_count":20,"threshold_duration":1440,"type":"runtime_error"}
terminate called after throwing an instance of 'std::runtime_error'
  what():  {"action":"reset","category":"hardware","code":"0xa35e","description":"The card encountered a problem and triggered a PCIe bus fence operation","message":"PCIe bus master fence","name":"RAS::PCI::BusFence","severity":"ERROR","step":"Reset, reseat, or replace the card","threshold_action":"deconfigure","threshold_count":20,"threshold_duration":1440,"type":"runtime_error"}
Signal Received: 6 (Aborted)
Signal Received from pid=121078
*** BACKTRACE ***
/opt/ibm/spyre/senlib/lib/libsenlib-dd2.so(+0x44e05f)[0x7efe6323a05f]
/lib64/libc.so.6(+0x411c0)[0x7eff21ba31c0]
/lib64/libc.so.6(+0x9781c)[0x7eff21bf981c]
/lib64/libc.so.6(gsignal+0x16)[0x7eff21ba3116]
/lib64/libc.so.6(abort+0xd7)[0x7eff21b8a8fa]
/lib64/libstdc++.so.6(+0xa603a)[0x7eff0a79203a]
/lib64/libstdc++.so.6(+0xbc0ac)[0x7eff0a7a80ac]
/lib64/libstdc++.so.6(_ZSt10unexpectedv+0x0)[0x7eff0a791a39]
/lib64/libstdc++.so.6(+0xbc338)[0x7eff0a7a8338]
/opt/ibm/spyre/senlib/lib/libsenlib-dd2.so(_ZN3RAS3PCI8BusFence5ThrowEv+0x64)[0x7efe634fe8d4]
/opt/ibm/spyre/senlib/lib/libsenlib-dd2.so(_ZN6senlib2v212PfMSIMonitor13PollingThreadEv+0x2b78)[0x7efe634fde98]
/lib64/libstdc++.so.6(+0xee0e4)[0x7eff0a7da0e4]
/lib64/libc.so.6(+0x958bf)[0x7eff21bf78bf]
/lib64/libc.so.6(__clone+0x44)[0x7eff21c68264]
*****************
```

**Comment by KanakaPathivada** [comment]
cc: @pradghos @shruthip-venkatesh @joyalbin
